### PR TITLE
[SPARK-19257][SQL]location for table/partition/database should be java.net.URI

### PR DESCRIPTION
**Pull request by windpiger**
## What changes were proposed in this pull request?

Currently we treat the location of table/partition/database as URI string.

It will be safer if we can make the type of location as java.net.URI. 

In this PR, there are following classes changes:
**1. CatalogDatabase**
```
case class CatalogDatabase(
    name: String,
    description: String,
    locationUri: String,
    properties: Map[String, String])
--->
case class CatalogDatabase(
    name: String,
    description: String,
    locationUri: URI,
    properties: Map[String, String])
```
**2. CatalogStorageFormat**
```
case class CatalogStorageFormat(
    locationUri: Option[String],
    inputFormat: Option[String],
    outputFormat: Option[String],
    serde: Option[String],
    compressed: Boolean,
    properties: Map[String, String])
---->
case class CatalogStorageFormat(
    locationUri: Option[URI],
    inputFormat: Option[String],
    outputFormat: Option[String],
    serde: Option[String],
    compressed: Boolean,
    properties: Map[String, String])
```

Before and After this PR, it is transparent for user, there is no change that the user should concern. The `String` to `URI` just happened in SparkSQL internally.

Here list some operation related location:
**1. whitespace in the location**
   e.g.  `/a/b c/d`
   For both table location and partition location, 
   After `CREATE TABLE  t... (PARTITIONED BY ...) LOCATION '/a/b c/d'` ,
   then `DESC EXTENDED t ` show the location is `/a/b c/d`,
   and the real path in the FileSystem also show `/a/b c/d`

**2. colon(:) in the location**
   e.g.  `/a/b:c/d`
   For both table location and partition location, 
   when `CREATE TABLE  t... (PARTITIONED BY ...)  LOCATION '/a/b:c/d'` ,
   
  **In linux file system**
   `DESC EXTENDED t ` show the location is `/a/b:c/d`,
   and the real path in the FileSystem also show `/a/b:c/d`

  **in HDFS** throw exception:
  `java.lang.IllegalArgumentException: Pathname /a/b:c/d from hdfs://iZbp1151s8hbnnwriekxdeZ:9000/a/b:c/d is not a valid DFS filename.`
 

  **while** After `INSERT INTO TABLE t PARTITION(a="a:b") SELECT 1`
   then `DESC EXTENDED t ` show the location is `/xxx/a=a%3Ab`,
   and the real path in the FileSystem also show `/xxx/a=a%3Ab`

**3. percent sign(%) in the location**
   e.g.  `/a/b%c/d`
   For both table location and partition location, 
   After `CREATE TABLE  t... (PARTITIONED BY ...) LOCATION '/a/b%c/d'` ,
   then `DESC EXTENDED t ` show the location is `/a/b%c/d`,
   and the real path in the FileSystem also show `/a/b%c/d`

**4. encoded(%25) in the location**
   e.g.  `/a/b%25c/d`
   For both table location and partition location, 
   After `CREATE TABLE  t... (PARTITIONED BY ...)  LOCATION '/a/b%25c/d'` ,
   then `DESC EXTENDED t ` show the location is `/a/b%25c/d`,
   and the real path in the FileSystem also show `/a/b%25c/d`

   **while** After `INSERT INTO TABLE t PARTITION(a="%25") SELECT 1`
   then `DESC EXTENDED t ` show the location is `/xxx/a=%2525`,
   and the real path in the FileSystem also show `/xxx/a=%2525`

**Additionally**, except the location, there are two other factors will affect the location of the table/partition. one is the table name which does not allowed to have special characters, and the  other is `partition name` which have the same actions with `partition value`, and `partition name` with special character situation has add some testcase and resolve a bug in [PR](https://github.com/apache/spark/pull/17173)

### Summary:
After `CREATE TABLE  t... (PARTITIONED BY ...)  LOCATION path`,
the path which we get from `DESC TABLE` and `real path in FileSystem` are all the same with the `CREATE TABLE` command(different filesystem has different action that allow what kind of special character to create the path, e.g. HDFS does not allow colon, but linux filesystem allow it ).

`DataBase` also have the same logic with `CREATE TABLE`

while if the `partition value` has some special character like `%` `:` `#` etc, then we will get the path with encoded `partition value` like `/xxx/a=A%25B` from `DESC TABLE` and `real path in FileSystem`

In this PR, the core change code is using `new Path(str).toUri` and `new Path(uri).toString`
which transfrom `str to uri `or `uri to str`.
for example: 
```
val str = '/a/b c/d'
val uri = new Path(str).toUri  --> '/a/b%20c/d'
val strFromUri = new Path(uri).toString -> '/a/b c/d'
```

when we restore table/partition from metastore, or get the location from `CREATE TABLE` command, we can use it as above to change string to uri `new Path(str).toUri `

## How was this patch tested?
unit test added.
The `current master branch` also `passed all the test cases` added in this PR by a litter change.
https://github.com/apache/spark/pull/17149/files#diff-b7094baa12601424a5d19cb930e3402fR1764
here `toURI` -> `toString` when test in master branch.

This can show that this PR  is transparent for user.

### CODE REVIEW

**Comment by windpiger** [comment]
cc @cloud-fan @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 1]
```diff
@@ -397,7 +398,8 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
     val tableDesc = CatalogTable(
       identifier = table,
       tableType = tableType,
-      storage = storage.copy(locationUri = customLocation),
+      storage = storage.copy(locationUri = customLocation.map{ loc =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: nit:
```
storage.copy(
  locationUri = xxx)
```

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: also add comments to explain why we don't create `URI` directly.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: this pattern appears many times in this PR, how about we create a util method `toURI` in `CatalogUtils` and add comments there?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 2]
```diff
@@ -1843,28 +1846,28 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
              |OPTIONS(path "$dir")
            """.stripMargin)
         val table = spark.sessionState.catalog.getTableMetadata(TableIdentifier("t"))
-        assert(table.location == dir.getAbsolutePath)
+        assert(table.location == new URI(dir.getAbsolutePath))
 
         dir.delete
-        val tableLocFile = new File(table.location)
-        assert(!tableLocFile.exists)
+//        val tableLocFile = new File(table.location)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: why comment it out?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 3]
```diff
@@ -565,7 +565,8 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
 
         val newLocation = tableDefinition.storage.locationUri
         val storageWithPathOption = tableDefinition.storage.copy(
-          properties = tableDefinition.storage.properties ++ newLocation.map("path" -> _))
+          properties =
+            tableDefinition.storage.properties ++newLocation.map("path" -> new Path(_).toString))
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: we should also add a util method `URIToString` and add comments

**Review by cloud-fan — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala [hunk 4]
```diff
@@ -883,7 +885,7 @@ abstract class CatalogTestUtils {
 
   def newFunc(): CatalogFunction = newFunc("funcName")
 
-  def newUriForDatabase(): String = Utils.createTempDir().toURI.toString.stripSuffix("/")
+  def newUriForDatabase(): URI = new URI(Utils.createTempDir().toURI.toString.stripSuffix("/"))
```
Comment (windpiger) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: `Utils.createTempDir().toURI` has a suffix '/', here we have to strip it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala [hunk 5]
```diff
@@ -162,6 +164,28 @@ object CatalogUtils {
     BucketSpec(numBuckets, normalizedBucketCols, normalizedSortCols)
   }
 
+  /**
+   * Convert URI to String.
+   * Since URI.toString does not decode for the uri string, we need to use
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala]: `Since URI.toString does not decode the uri string, here we create a hadoop Path with the given URI, ad rely on hadoop Path.toString to decode the uri`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala [hunk 6]
```diff
@@ -162,6 +164,28 @@ object CatalogUtils {
...
+   * Path(uri).toString to decode it.
+   * @param uri the URI of the path
+   * @return the String of the path
+   */
+  def URIToString(uri: Option[URI]): Option[String] = {
+    uri.map(new Path(_).toString)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala]: can you try `java.net.URLDecoder`? I'd like to avoid hadoop dependency on this kind of basic functionality.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 7]
```diff
@@ -131,7 +132,7 @@ class SessionCatalog(
    * does not contain a scheme, this path will not be changed after the default
    * FileSystem is changed.
    */
-  private def makeQualifiedPath(path: String): Path = {
+  private def makeQualifiedPath(path: URI): Path = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: this method can return `URI` too

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala (hunk 4 above)
Comment (windpiger) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: @cloud-fan  after I try to use Path to Compare , I think stripSuffix here is the simple way.

Path with a `String` type constructor will be equal when one has a `/`, and another does not.
```
scala> val x = new Path("/ab/c/")
x: org.apache.hadoop.fs.Path = /ab/c

scala> val y = new Path("/ab/c")
y: org.apache.hadoop.fs.Path = /ab/c

scala> x == y
res0: Boolean = true
```

Path with a `URI` type constructor will be not equal when one has a `/`, and another does not.
```
scala> val x =new URI("/a/b/c/")
x: java.net.URI = /a/b/c/

scala> val y =new URI("/a/b/c")
y: java.net.URI = /a/b/c

scala> x == y
res1: Boolean = false

scala> val x1 =new Path(x)
x1: org.apache.hadoop.fs.Path = /a/b/c/

scala> val y1 =new Path(y)
y1: org.apache.hadoop.fs.Path = /a/b/c

scala> x1 == y1
res2: Boolean = false
```

So just the Path with `String` type can ignore the suffix `/`, then if we have a `URI` in hand, and we want to compare with another `URI`, we should  first transform them to `String` , and use this String to constructor a Path, after this two actions, we can compare them with ignore the suffix `/`.

But I think it is more complicate, here we normalize the URI with stripSuffix from the Orignal then compare two URI directly, it is more simple.

should we must to convert it to Path to compare?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala [hunk 8]
```diff
@@ -553,7 +555,7 @@ abstract class ExternalCatalogSuite extends SparkFunSuite with BeforeAndAfterEac
   test("alter partitions") {
     val catalog = newBasicCatalog()
     try {
-      val newLocation = newUriForDatabase()
+      val newLocation = new Path(newUriForDatabase()).toUri
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: unnecessary change?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: I think it's ok if we always compare URI with Path, instead of converting it to string.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 9]
```diff
@@ -21,6 +21,7 @@ import scala.collection.JavaConverters._
 
 import org.antlr.v4.runtime.{ParserRuleContext, Token}
 import org.antlr.v4.runtime.tree.TerminalNode
+import org.apache.hadoop.fs.Path
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: we can remove this

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 10]
```diff
@@ -146,7 +150,7 @@ case class CreateDataSourceTableAsSelectCommand(
       assert(table.schema.isEmpty)
 
       val tableLocation = if (table.tableType == CatalogTableType.MANAGED) {
-        Some(sessionState.catalog.defaultTablePath(table.identifier))
+        Some(new Path(sessionState.catalog.defaultTablePath(table.identifier)).toUri)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: the `defaultTablePath` should also return URI now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 11]
```diff
@@ -146,7 +146,7 @@ case class DescribeDatabaseCommand(
     val result =
       Row("Database Name", dbMetadata.name) ::
         Row("Description", dbMetadata.description) ::
-        Row("Location", dbMetadata.locationUri) :: Nil
+        Row("Location", new Path(dbMetadata.locationUri).toString) :: Nil
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: use `CatalogUtil.URIToString`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 12]
```diff
@@ -66,7 +66,7 @@ case class CreateDatabaseCommand(
       CatalogDatabase(
         databaseName,
         comment.getOrElse(""),
-        path.getOrElse(catalog.getDefaultDBPath(databaseName)),
+        new Path(path.getOrElse(catalog.getDefaultDBPath(databaseName))).toUri,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: `getDefaultDBPath` should return URI now.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 13]
```diff
@@ -77,7 +79,7 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
     new Database(
       name = metadata.name,
       description = metadata.description,
-      locationUri = metadata.locationUri)
+      locationUri = new Path(metadata.locationUri).toString)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: URIToString?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 14]
```diff
@@ -95,7 +96,7 @@ private[sql] class SharedState(val sparkContext: SparkContext) extends Logging {
   // Create the default database if it doesn't exist.
   {
     val defaultDbDefinition = CatalogDatabase(
-      SessionCatalog.DEFAULT_DATABASE, "default database", warehousePath, Map())
+      SessionCatalog.DEFAULT_DATABASE, "default database", new Path(warehousePath).toUri, Map())
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala]: URIToString?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 15]
```diff
@@ -72,14 +73,14 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
 
   private def createDatabase(catalog: SessionCatalog, name: String): Unit = {
     catalog.createDatabase(
-      CatalogDatabase(name, "", spark.sessionState.conf.warehousePath, Map()),
+      CatalogDatabase(name, "", new Path(spark.sessionState.conf.warehousePath).toUri, Map()),
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: StringToURI?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 16]
```diff
@@ -72,14 +73,14 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
...
       ignoreIfExists = false)
   }
 
   private def generateTable(catalog: SessionCatalog, name: TableIdentifier): CatalogTable = {
     val storage =
       CatalogStorageFormat(
-        locationUri = Some(catalog.defaultTablePath(name)),
+        locationUri = Some(new Path(catalog.defaultTablePath(name)).toUri),
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: StringToURI?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 17]
```diff
@@ -146,7 +147,7 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
     try {
       sql(s"CREATE DATABASE $dbName")
       val db1 = catalog.getDatabaseMetadata(dbName)
-      val expectedLocation = makeQualifiedPath(s"spark-warehouse/$dbName.db")
+      val expectedLocation = new Path(makeQualifiedPath(s"spark-warehouse/$dbName.db")).toUri
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: StringToURI?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 18]
```diff
@@ -169,7 +170,7 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
           val dbNameWithoutBackTicks = cleanIdentifier(dbName)
           sql(s"CREATE DATABASE $dbName Location '$path'")
           val db1 = catalog.getDatabaseMetadata(dbNameWithoutBackTicks)
-          val expPath = makeQualifiedPath(tmpDir.toString)
+          val expPath = new Path(makeQualifiedPath(tmpDir.toString)).toUri
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: StringToURI?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 17 above)
Comment (windpiger) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: I just implement this function with a option[String], I think for a String without a option, it is already simple, we still to implement one?

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: I think `StringToURI` should convert `String` to `URI`, not `Option[String]`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala (hunk 6 above)
Comment (windpiger) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala]: what about StringToURI? it still import Path

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 17 above)
Comment (windpiger) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: ok, maybe we should implement them all? option[string] also exists in many places, and more complicate

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 19]
```diff
@@ -386,7 +386,7 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
         "LOCATION and 'path' in OPTIONS are both used to indicate the custom table path, " +
           "you can only specify one of them.", ctx)
     }
-    val customLocation = storage.locationUri.orElse(location)
+    val customLocation = storage.locationUri.orElse(location.map(CatalogUtils.stringToURI(_)))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: nit: `xx.map(CatalogUtils.stringToURI)`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 20]
```diff
@@ -1686,4 +1689,162 @@ class HiveDDLSuite
       }
     }
   }
+
+  Seq("a b", "a:b", "a%b").foreach { specialChars =>
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: please add a TODO to remove these duplicated tests when we merge DDLSuite and HiveDDLSuite

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 20 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: This introduces the conflict in my PR https://github.com/apache/spark/pull/16592. Let me remove the duplicate there.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 19 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: It seems this has a double-de/encoding problem when the input path is an URI.

```scala
scala> new org.apache.hadoop.fs.Path(new java.io.File("a b").toURI.toString).toUri.toString
res1: String = file:/.../a%2520b
```

A space character in URI is encoded as `%20` and `%` character is encoded as `%25`. A URI already has a %20 in it, and gets urlencoded again, from `%20` to `%2520`.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Also, it seems we have a similar util with `CatalogUtils.stringToURI` in `org.apache.spark.util.Utils.resolveURI`. Could we consolidate them? I did not replace it when I identified this because some existing tests in `org.apache.spark.util.UtilsSuite` were broken but I guess it would be fine if there is a coherent reason. These broken cases might be bugs.

@windpiger could you double check my comments and open a followup if I was not wrong?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 21]
```diff
@@ -386,7 +386,7 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: It seems this has a double-de/encoding problem when the input path is an URI.

```scala
scala> new org.apache.hadoop.fs.Path(new java.io.File("a b").toURI.toString).toUri.toString
res1: String = file:/.../a%2520b
```

A space character in URI is encoded as `%20` and `%` character is encoded as `%25`. A URI already has a %20 in it, and gets urlencoded again, from `%20` to `%2520`.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Also, it seems we have a similar util with `CatalogUtils.stringToURI` in `org.apache.spark.util.Utils.resolveURI`. Could we consolidate them? I did not replace it when I identified this because some existing tests in `org.apache.spark.util.UtilsSuite` were broken but I guess it would be fine if there is a coherent reason. These broken cases might be bugs.

@windpiger could you double check my comments and open a followup if I was not wrong?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: > I did not replace it when I identified this because some existing tests in org.apache.spark.util.UtilsSuite were broken but I guess it would be fine if there is a coherent reason. These broken cases might be bugs.

FYI.. cc @sarutak

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: If you reproduce it as a bug, please submit a PR to fix it.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: how could the input path be a URI string?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Hm.. is the input path always expected to be a path? I thought we support both URI and path forms.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: It's ambiguous to support both, what if users do wanna create a path `/tmp/%25`?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Up to my knowledge, HDFS's fully qualified path is a URI form. If we do not support this, that virtually means we are going to disallow the fully qualified path. I understand it sounds ambiguous but disallowing does not look a good solution.

Also, if users might want to access to local files or S3 when default scheme is `hdfs`, I guess it requires changing the default scheme every time.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: In addition, I guess we alway have a lot of tests with URI input paths and I did many of them to pass the tests on Windows, which I guess implicitly some committers do not disagree with this.

IMHO, I guess URIs should be supported first correctly and local path should be supported as abbreviation because those local path form is abbreviation of the fully qualified path.

Comment (sarutak) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: > how could the input path be a URI string?
How about external table on S3?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 22]
```diff
@@ -285,7 +285,7 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
         // compatible format, which means the data source is file-based and must have a `path`.
         require(table.storage.locationUri.isDefined,
           "External file-based data source table must have a `path` entry in storage properties.")
-        Some(new Path(table.location).toUri.toString)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: @HyukjinKwon previously we also encode the table location string, I think we never supported uri path string before?

Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: I see, it seems the problem itself seems existing before.

I found this while running related tests on Windows which it looks related with this PR (special character stuff) so I think this is a proper JIRA to make a followup with and PR to note this.

It seems we do have tests that use URIs already whether it was mistakenly supported or not. Should we ask it to dev-mailing list? I think this is an important decision.

**Comment by cloud-fan** [comment]
yea, please go ahead

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala (hunk 22 above)
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: yea, please go ahead

Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Thank you for confirming. Let me send a mail soon.

Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Actually, I did a bit of search more about what `org.apache.hadoop.fs.Path` expects. It seems the [documentation](https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/fs/Path.html) says:

> Construct a path from a String. Path strings are URIs, but with unescaped elements and some additional normalization.

It seems those strings are expected to be unescaped. So, it seems we support URI with unescaped characters, which is inherited from `Path`.

I want to be sure on this because I have fixed many tests to use URIs to pass on Windows and I am about to fix them further in this way. @steveloughran, do you mind if I cc you and ask to take a look and help to confirm this please? I know no one who knows better about Hadoop.

Comment (steveloughran) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: aah, paths, especially on windows make me weep. Note also lots of fun related to encoded secrets in s3n/s3a/s3 urls (HADOOP-3733 and the associated regression HADOOP-14114)

Because of those patches, and the fact that if you want to embed a / in an S3A URI you *must* escape it, I don't know what to say here.

If you can come up with a simple set of questions to ask someone who doesn't know this code but could understand Path's logic, I could pass them on to some people I think may know the answer. Maybe @Cnauroth might have some insight on what windows gets up to

BTW, Path in Hadoop 2.9 is Serializable. you can pass them down to closures working on RDDs, etc.

Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Thank you so much @steveloughran. Let me try to produce a simple set of questions purely about `org.apache.hadoop.fs.Path`.

Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: I think I have two questions as below:

- Is it okay to use both URIs and local file paths for the input string for `org.apache.hadoop.fs.Path` in general (when they are expected to be unescaped)?

- What `org.apache.hadoop.fs.Path` expects for input string? It seems URI strings but with unescaped characters per https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/fs/Path.html and there are special cases that requires escaped characters such as s3n/s3a/s3 urls. Is my understanding correct?

Comment (steveloughran) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: there's actually some discussion [about defining this properly](https://issues.apache.org/jira/browse/HADOOP-14217), related to the problem of "colons in object store paths". That's not going to help directly, but a sign of something which has historically been underspecified

Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Thank you. Your answer and the pointer helped me a lot to understand. 

So, up to my understanding...

There are few exceptional cases, for example, special characters or encoded characters in those URIs. Also, there are some corner cases that are not clearly defined about what `Path` expects.

Nevertheless, the baseline is fully qualified URI with unescaped characters or the unescaped path part in URI as the documentation says but there may be some variants.

**Comment by cnauroth** [comment]
@HyukjinKwon , nice to meet you!  I see I got notified here for a bit of Hadoop `Path` knowledge, and particularly on Windows.

> Is it okay to use both URIs and local file paths for the input string for org.apache.hadoop.fs.Path in general (when they are expected to be unescaped)?

Yes, this is correct.

Specifically on the topic of Windows, `Path` has special case logic for handling a Windows-specific local file path.  (This logic is only triggered if it detects the runtime OS is Windows.)  On Windows, I expect a call like `new Path("C:\\foo\\bar").toUri` to yield a correct `URI` pointing at that local file path, and further calling `toString` yields a correct `String` representation of the path.  Hadoop code often needs to take a path string that is possibly a relative path and pass it through `Path` to make it absolute and escape it according to Hadoop code expectations.

The standard invocation for doing this in the Hadoop code is `new Path(...).toUri();` or `new Path(...).toUri().toString();`.  This works across all platforms.  I don't have any knowledge of the Spark codebase, but I see this patch uses similar invocations, so I expect it's good.

**Comment by HyukjinKwon** [comment]
@cnauroth Thank you so much for your help.

**Comment by HyukjinKwon** [comment]
(I think I should cc @srowen too FYI because he reviewed all my PRs fixing the tests on Windows. The point here seems to me okay to use URIs in general.)

**Comment by gatorsmile** [comment]
Our parser might need a change regarding escape handling. We are having a related discussion in another PR: https://github.com/apache/spark/pull/15398

**Comment by HyukjinKwon** [comment]
@gatorsmile, Thanks for your pointer. There is a good discussion there.
